# NumPy Calculation Methods 

An array has various methods that perform calculations using its contents. By default, these methods ignore the array’s shape and use all the elements in the calculations. For example, calculating the mean of an array totals all of its elements regardless of its shape, then divides by the total number of elements. You can perform these calculations on each dimension as well. For example, in a two-dimensional array, you can calculate each row’s mean and each column’s mean. Consider an array representing four students’ grades on three exams:

In [1]:
import numpy as np 
grades = np.array([[97, 86, 72], [100, 87, 90],[94, 77, 90], [100, 81, 82]])
grades

array([[ 97,  86,  72],
       [100,  87,  90],
       [ 94,  77,  90],
       [100,  81,  82]])

We can use methods to calculate <span style="color: green;">sum, min, max, mean, std (standard deviation),var (variance)</span> 

In [2]:
grades.sum()

1056

In [3]:
grades.min()

72

In [4]:
grades.max()

100

In [5]:
grades.mean()

88.0

In [6]:
grades.std()

8.563488385776752

In [7]:
grades.var()

73.33333333333333

## Calculations by Row or Column
Many calculation methods can be performed on specific array dimensions, known as the array’s axes. These methods receive an axis keyword argument that specifies which dimension to use in the calculation, giving you a quick way to perform calculations by row or column in a two-dimensional array. Assume that you want to calculate the average grade on each exam, represented by the columns of grades. Specifying <span style="color: red;">axis=0 </span> performs the calculation on all the row values within each column:

In [8]:
grades.mean(axis=0)

array([97.75, 82.75, 83.5 ])

So 97.75 above is the average of the first column’s grades (97, 100, 94 and 100), 82.75 is the average of the second column’s grades and 83.5 is the average of the third column’s grades. 

Similarly, specifying <span style="color: red;">axis=1</span>  performs the calculation on all the column values within each individual row. To calculate each student’s average grade for all exams, we can use:

In [9]:
grades.mean(axis=1)

array([85.        , 92.33333333, 87.        , 87.66666667])

This produces four averages—one each for the values in each row. So 85. is the average of row 0’s grades, and the other averages are for the remaining rows. NumPy arrays have many more calculation methods. For the complete list, see <span style="color: red;">https://docs.scipy.org/doc/numpy/reference/arrays.ndarray.html</span> 